Tests/AsyncQueueTests/SemaphoreTests.swift

```diff
@@ -0,0 +1,126 @@
+// MIT License
+//
+// Copyright (c) 2022 Dan Federman
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+import XCTest
+
+@testable import AsyncQueue
+
+final class SemaphoreTests: XCTestCase {
+
+    // MARK: XCTestCase
+
+    override func setUp() async throws {
+        try await super.setUp()
+
+        systemUnderTest = Semaphore()
+    }
+
+    override func tearDown() async throws {
+        try await super.tearDown()
+
+        while await systemUnderTest.isWaiting {
+            await systemUnderTest.signal()
+        }
+    }
+
+    // MARK: Behavior Tests
+
+    func test_wait_suspendsUntilEqualNumberOfSignalCalls() async {
+        actor CountingExecutor {
+            func enqueue(andCount incrementOnCompletion: Bool = false, _ task: @escaping @Sendable () async -> Void) async {
+                await withCheckedContinuation { continuation in
+                    // Re-enter the actor context but don't wait for the result.
+                    Task {
+                        // Now that we're back in the actor context, resume.
+                        continuation.resume()
```

Comment:
I'm having trouble following the continuation -> task dance that we're doing here. Can you help me follow why that's necessary? When `enqueue(…)` starts executing we're already on the actor context so why is it necessary to create a continuation and then create a task to get back on the actor context?

Reply:
Overall I think an explanation would be helpful for how the inlined actor improved this test. Based on https://github.com/dfed/swift-async-queue/pull/3#discussion_r1041229725 I see that this is definitely an improvement, though I'd love to better understand the logic/reasoning.

Reply:
I'll add some comments in the next pass. Short version: we're using a trick similar to the `ActorQueue` to enable ordered execution of tasks from a single context.

The problems I was solving for:
1. I needed to ensure we are calling `wait()` before `signal()`
2. I needed to not `await` the `wait()` call, since that would have hung the test

In order to do that, I used an `Actor` (which has inherit task ordering) to serialize enqueued tasks and ensure the `wait()` call would suspend because it was executing before we `signal()`'d.

Reply:
Next pass comments: d1a5560